faststreams/output_stream.nim

```diff
@@ -369,25 +431,27 @@ proc append*(c: var WriteCursor, chars: openarray[char]) {.inline.} =
   var charsStart = unsafeAddr chars[0]
   c.append makeOpenArray(cast[ptr byte](charsStart), chars.len)
 
-template appendMemCopy*(c: var WriteCursor, value: auto) =
+template appendMemCopy*[T](c: var WriteCursor, value: T) =
   bind append
-  # TODO: add a check that this is a trivial type
+  static:
+    type TT = T # TODO This deals with a Nim bug
+    assert supportsCopyMem(TT)
```

Comment:
this looks like something that should be bundled with the byte-casting and the open-array making so there is one "safe:ish" template for preparing a memcpy - also, what's common in C++ for example is that this happens automatically so that memcpy:able types get the memcpy treatment while the others get normal assignment semantics, removing the failure mode for the caller and doing the best possible thing for the given type always

Reply:
`supportsCopyMem` was introduced in Nim 1.2, so I did this minimal safety enhancement to the existing API. I need to study the implementation of `supportsCopyMem` much more carefully before we can deem it safe for application in the more general `append` case. 

Please note that `append` doesn't allow you to pass any type of objects at the moment. Having something like this would be a bit undesirable when you add `nim-serialization` to the picture, because appending an object should always also specify the serialization scheme being used to encode it.

Reply:
oh my, and `supportsCopyMem` was chosen as a name? that smells of a feature that was just thrown in for a narrow use case without considering the bigger picture :/

Reply:
I've been using `supportsCopyMem` for months. What changed in 1.0 was that types with custom destructors had `supportsCopyMem` set to false
![image](https://user-images.githubusercontent.com/22738317/79754396-c4f37f00-8317-11ea-960b-32f1eb6594f7.png)

It's used in the compiler for deepCopy implementation and other proc that care about trivial types or ref/seq/string/destructors